Add support to supply your binary/options.

thakyZ opened the pull request:
- Add support to supply own binary.
- Add support to specify custom user to download from.
- Refactored some code.

The script falls back to downloading if unsuccessful.   
Maybe add an option to not fall back to downloading, but that can also be easily achieved via setting the environment variable `TSGO_PROVIDER` to an empty string.

Added environment variables:   
`TSGO_DIRECTORY`: The directory, where the `typescript-go` project is, should contain the partial path `./built/local/tsgo.exe`.
`TSGO_PROVIDER`: A GitHub username that has releases for `tsgo`

thakyZ committed:
update: add support to supply your options.
- add support to supply own binary
- add support to specify custom user to download from.
- refactored some code.

rxliuli commented:
> TSGO_DIRECTORY

Perhaps a better implementation would be some kind of global cache? First, automatically download to a global location cache, and then if a cached version is found, automatically copy it.

> TSGO_PROVIDER

Of course, this option looks good, and it would be great if you could add it to the README.

thakyZ commented:
> > TSGO_DIRECTORY
> 
> Perhaps a better implementation would be some kind of global cache? First, automatically download to a global location cache, and then if a cached version is found, automatically copy it.

The point of the `TSGO_DIRECTORY` is to be able to supply your own build, in case you do not want to download a copy someone else built.

rxliuli commented:
> The point of the `TSGO_DIRECTORY` is to be able to supply your own build, in case you do not want to download a copy someone else built.

Oh, that's right. If you could add the descriptions of these two environment variables to the README, it should help others better understand them.

rxliuli commented on install.js:
```diff
@@ -38,6 +45,24 @@ if (existsSync(binaryPath)) {
 
 async function download() {
+  if (env.TSGO_DIRECTORY) {
+    const filePath = join(env.TSGO_DIRECTORY, 'built', 'local', 'tsgo.exe')
```
Strangely, the binary files for mac/linux do not end with `.exe`. You should use binaryName, which calculates the correct file name based on the platform.

thakyZ replied:
Whoops, almost forgot about that. Thanks for pointing it out. I have made changes and pushed them.

thakyZ committed:
update: changed to fit requests.
- add environemnt variables to README.md
- accounting for someone not installing tsgo from a local windows copy

rxliuli reviewed (approved):
LGTM. it will be automatically applied when the new version is released tomorrow after merging.
